### PR TITLE
test(datepicker): correctly treat TestBed.compileComponents as async

### PR DESCRIPTION
**Pull request by jelbourn**
Tests for datepicker previously assumed that Test.compileComponents
completes synchronously, immediately instantiating components after
calling the compile. This is not true, and the tests only work by
coicidence. With ivy, this is not as lenient and starts failing.

### CODE REVIEW

File: src/lib/datepicker/datepicker.spec.ts [hunk 1]
```diff
@@ -25,50 +33,95 @@ import {MatDatepickerInput} from './datepicker-input';
...
+  let scrolledSubject: Subject<void>;
+
+  /** Fake `Directionality` service. */
+  let fakeDirectectionality: Partial<Directionality>;
+
+  /** Fake `change` stream for the fake `Directionality` service */
+  let fakeDirectionChanges: EventEmitter<Direction>;
+
+  // Always clean up the overlay container after every test.
+  afterEach(inject([OverlayContainer], (container: OverlayContainer) => {
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.spec.ts]: nit: move after the `beforeEach`

File: src/lib/datepicker/datepicker.spec.ts [hunk 2]
```diff
@@ -25,50 +33,95 @@ import {MatDatepickerInput} from './datepicker-input';
...
-  function createComponent(
-    component: Type<any>,
-    imports: Type<any>[] = [],
-    providers: (FactoryProvider | ValueProvider)[] = [],
-    entryComponents: Type<any>[] = []): ComponentFixture<any> {
+  /**  Fake subject for signaling that a scroll has occurred. */
+  let scrolledSubject: Subject<void>;
+
+  /** Fake `Directionality` service. */
+  let fakeDirectectionality: Partial<Directionality>;
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.spec.ts]: typo (ectect)

File: src/lib/datepicker/datepicker.spec.ts [hunk 3]
```diff
@@ -25,50 +33,95 @@ import {MatDatepickerInput} from './datepicker-input';
...
+        commonNgModuleImports,
+        MatNativeDateModule,
       ],
-      providers,
-      declarations: [component, ...entryComponents],
-    });
-
-    TestBed.overrideModule(BrowserDynamicTestingModule, {
+      declarations: [
+        StandardDatepicker,
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.spec.ts]: I think we decided that compiling all of the components was making our tests super slow, that's why we switched from something like this to what we currently have

**Comment by andrewseguin** [comment]
Needs rebase

**Pull request closed without merging**